lib/common/socket.c

```diff
@@ -960,6 +991,20 @@ void h2o_socket_ssl_handshake(h2o_socket_t *sock, SSL_CTX *ssl_ctx, const char *
         else
             h2o_socket_read_start(sock, proceed_handshake);
     } else {
+        if (session_cache != NULL) {
+            sock->ssl->handshake.client.session_cache = session_cache;
+            h2o_iovec_t session_cache_key = get_ssl_session_cache_key(sock, server_name);
+            sock->ssl->handshake.client.session_cache_key = session_cache_key;
+            sock->ssl->handshake.client.session_cache_key_hash = h2o_cache_calchash(session_cache_key.base, session_cache_key.len);
+            fprintf(stderr, "@@@ session_cache_key: %s\n", session_cache_key.base); // FIXME: nagata
+
+            h2o_cache_ref_t *cacheref =
+                h2o_cache_fetch(session_cache, h2o_now(h2o_socket_get_loop(sock)), sock->ssl->handshake.client.session_cache_key,
+                                sock->ssl->handshake.client.session_cache_key_hash);
+            if (cacheref != NULL) {
+                SSL_set_session(sock->ssl->ssl, (SSL_SESSION *)cacheref->value.base);
```

Comment:
A non-NULL reference returned by `h2o_cache_fetch` should be released by calling `h2o_cache_release`.